README.md
# digital-navy

Run lots of fully isolated commands on digital ocean

[![Build Status](https://img.shields.io/travis/ForbesLindesay/digital-navy/master.svg)](https://travis-ci.org/ForbesLindesay/digital-navy)
[![Dependency Status](https://img.shields.io/david/ForbesLindesay/digital-navy/master.svg)](http://david-dm.org/ForbesLindesay/digital-navy)
[![NPM version](https://img.shields.io/npm/v/digital-navy.svg)](https://www.npmjs.org/package/digital-navy)

## Installation

```
npm install digital-navy --save
```

## Usage

```js
var fs = require('fs');
var DigitalNavy = require('digital-navy');

var script = fs.readFileSync(__dirname + '/example-script.js');

var navy = new DigitalNavy({
  // Note that the base image is only re-built when the name changes,
  // so we include a shasum of the script and the node version
  name: 'node-' + process.version + '-' + shasum(script),
  // the prepare function is used to run ssh commands before creating
  // the snapshot.  Use it to install all your dependencies
  prepare(ssh) {
    console.log('installing node');
    // we install the same version of node as we're using locally
    return DigitalNavy.installNode(ssh, process.version).then(
      () => {
        console.log('installed node');
        // copy some JavaScript script to the server
        return ssh.exec('cat > ~/example-script.js', {stdin: script});
      }
    );
  },
  token: '<DIGITAL OCEAN KEY>',
  // keypair generated by either rsa-json or keypair
  keypair: keypair(),

  // You can use functions for minPoolSize and maxPoolSize if you want to have these change over time
  // changes in the return values of those functions will only be taken into
  // account when droplets start or stop executing
  minPoolSize: 3,
  maxPoolSize: 4,
});

// Call this function each time you want to run some commands in their own isolated sandbox
navy.run(
  ssh => {
    console.log('got droplet');
    return ssh.exec('node ~/example-script.js').then(
      result => {
        console.log('finished script');
        console.dir(result);
        assert(result === 'Hello World\n');
        return result;
      },
    );
  }
);
```

### DigitalNavy constructor

options:

name        | default         | description
------------| --------------- |-----------------
name        | **required**    | The name for the snapshot and instances.  The snapshot is rebuilt whenever this name changes, so use it to encode all versioning info.
prepare     | **required**    | A function to be called with an ssh connection to the droplet that will be used to generate a snapshot.  Use it to install dependencies.
token       | **required**    | A digital ocean token, used to create the droplets 😊
keypair     | **required**    | An object with `{public, private}` keys.  Generated with [keypair](https://www.npmjs.com/package/keypair) or [rsa-json](https://www.npmjs.com/package/rsa-json)
maxPoolSize | `2`             | The maximum number of isolated virtual machines to allocate at a time.  By default this is kept low to prevent you accidentally spending lots of money.
minPoolSize | `0`             | The minimum number of workers to keep running at a time.  Keeping servers running even when there's no work to do helps a lot with latency.
workerSize  | `'512mb'`       | The size of the droplet allocated for performing the work.  Increase this if you need more memory or CPU for each worker.
builderSize | `= workerSize`  | The size of the droplet used to build the initial sandbox.  Use a larger droplet if you need to build big assets up front, but the actual worker droplets can be much smaller.


### digitalNavy.run(options, fn) _(instance method)_

Run a sequence of ssh commands in an isolated droplet.  The function `fn` is given an instance of `SshConnection`.

options:

name        | default         | description
------------| --------------- |-----------------
debug       | `false`         | Set this to `true` to emit the stdout & stderr or remote commands on the local terminal

### DigitalNavy.installNode(ssh, version) _(static method)_

A helper method to install a chosen version of node.js via an `SshConnection`.  You probably want to use this in your `prepare` function.


### SshConnection#exec(commands, options) _(instance method)_

execute a command or sequence of commands and return a `Promise` for the string value of `stdout`.

name        | default                 | description
------------| ----------------------- |-----------------
debug       | `= parentOptions.debug` | Set this to `true` to override the default value set in `digitalNavy.run`
stdin       | null                    | Set this to a Buffer or string to provide it as the stdin for the command being run

## License

MIT


HISTORY.md
# Changelog

## v0.0.1: 2016-16-19

- Initial release


test/index.js
import {readFileSync, writeFileSync} from 'fs';
import assert from 'assert';
import Promise from 'promise';
import shasum from 'shasum';
import keypair from 'keypair';
import './ssh-mock';
import {_disableTimeouts} from '../src/digital-ocean';
import DigitalNavy, {installNode} from '../src';

_disableTimeouts();
DigitalNavy._mockDate(() => '2016-06-19T01:39:40.763Z');

const script = readFileSync(__dirname + '/example-script.js');

let keys = null;
if (process.env.DIGITAL_OCEAN_KEY) {
  keys = keypair();
  writeFileSync(__dirname + '/public-key', keys.public);
} else {
  keys = {
    public: readFileSync(__dirname + '/public-key', 'utf8'),
    // the private key isn't really used when running against mocks so doesn't need to match public key
    private: keypair().private,
  };
}
const navy = new DigitalNavy({
  // Note that the base image is only re-built when the name changes, so we include a shasum of the script and the node
  // version
  name: 'node-' + process.version + '-' + shasum(script),
  prepare(ssh) {
    console.log('installing node');
    // we install the same version of node as we're using locally
    return installNode(ssh, process.version).then(
      () => {
        console.log('installed node');
        // copy the script to the server
        return ssh.exec('cat > ~/example-script.js', {stdin: script});
      }
    );
  },
  token: process.env.DIGITAL_OCEAN_KEY || 'DIGITAL_OCEAN_KEY',
  keypair: keys,
  // You can use functions here if you want to have these change over time
  // changes in the return values of those functions will only be taken into
  // account when droplets start or stop executing
  minPoolSize: 3,
  maxPoolSize: 4,
});

function run() {
  return navy.run(
    ssh => {
      console.log('got droplet');
      return ssh.exec('node ~/example-script.js').then(
        result => {
          console.log('finished script');
          console.dir(result);
          assert(result === 'Hello World\n');
          return result;
        },
      );
    }
  );
}

// You don't have to wait for the navy to be ready, but I do it here so that we can see timing for a fully initialized
// pool of workers
navy.ready.done(() => {
  const start = Date.now();
  Promise.all([run(), run(), run(), run(), run()]).done(
    result => {
      const end = Date.now();
      console.dir(result);
      console.log('Script executed in ' + ((end - start) / 1000) + ' seconds');
    }
  );
});


test/ssh-mock.js
import assert from 'assert';
import {readFileSync, writeFileSync, mkdirSync} from 'fs';
import rimraf from 'rimraf';
import Promise from 'promise';
import {stringify, parse} from 'json-buffer';
import DigitalNavy from '../src/index';
import SSH from '../src/ssh';

if (process.env.DIGITAL_OCEAN_KEY) {
  rimraf.sync(__dirname + '/recorded-ssh');
  mkdirSync(__dirname + '/recorded-ssh');
}

class SSHRecorder {
  constructor(remote, options) {
    this._id = remote.host;
    const args = [{host: remote.host, port: remote.port, username: remote.port}, options];
    this._record = {
      constructor: args,
      execs: [],
    };
    this._client = new SSH(remote, options);
    this.ready = this._client.ready;
  }

  exec(...args) {
    const action = {args};
    this._record.execs.push(action);
    const result = this._client.exec(...args);
    result.done(result => action.result = result);
    return result;
  }

  close() {
    setTimeout(() => {
      writeFileSync(__dirname + '/recorded-ssh/' + this._id, stringify(this._record));
    }, 0);
    return this._client.close();
  }
}

class SSHReplayer {
  constructor(remote, options) {
    this._id = remote.host;
    const args = [{host: remote.host, port: remote.port, username: remote.port}, options];
    this._record = parse(readFileSync(__dirname + '/recorded-ssh/' + this._id, 'utf8'));
    assert.deepEqual(parse(stringify(args)), this._record.constructor);
    this._index = 0;
    this.ready = Promise.resolve(null);
  }

  exec(...args) {
    const call = this._record.execs[this._index++];
    assert.deepEqual(parse(stringify(args)), call.args);
    return Promise.resolve(call.result);
  }

  close() {
    assert.strictEqual(this._index, this._record.execs.length);
    return Promise.resolve(null);
  }
}
DigitalNavy._mockSsh(process.env.DIGITAL_OCEAN_KEY ? SSHRecorder : SSHReplayer);


test/example-script.js
const fs = require('fs');

try {
  fs.readFileSync(__dirname + '/run-artifact');
  throw new Error('Environment not properly cleaned up');
} catch (ex) {
  if (ex.code !== 'ENOENT') throw ex;
  fs.writeFileSync(__dirname + '/run-artifact', 'this script should always be run in a fresh build');
}

console.log('Hello World');


src/ssh.js
import Promise from 'promise';
import SSHConnection from 'ssh2';


// N.B. this file is completely un-tested when unless you provide a value for process.env.DIGITAL_OCEAN_KEY

export default class SSH {
  constructor(remote, options) {
    this._options = options || {};
    this.ready = new Promise((resolve, reject) => {
      const attempt = (attemptNumber) => {
        this._connection = new SSHConnection();
        this._connection.on('error', err => {
          if (attemptNumber < 5) {
            console.log('connection failed, retrying in ' + Math.pow(2, attemptNumber) + ' second(s)');
            setTimeout(() => attempt(attemptNumber + 1), 1000 * Math.pow(2, attemptNumber));
          } else {
            reject(err);
          }
        });
        this._connection.on('ready', resolve);
        this._connection.connect(remote);
      };
      attempt(0);
    });
  }

  exec(commands, options = {}) {
    if (Array.isArray(commands)) commands = commands.join(' && ');
    const debug = (options.debug === undefined ? this._options.debug : options.debug) || false;

    if (debug) {
      console.log('ssh-exec: ' + commands);
    }

    const result = this.ready.then(() => {
      return new Promise((resolve, reject) => {
        let exitCode = 1;
        this._connection.exec(commands, {pty: options.pty}, (err, stream) => {
          if (err) return reject(err);
          let stream_output = '';
          let err_output = '';
          stream.on('data', (data) => {
            stream_output += data.toString();
            if (debug) {
              process.stdout.write(data);
            }
          });
          stream.stderr.on('data', (data) => {
            err_output += data.toString();
            if (debug) {
              process.stderr.write(data);
            }
          });
          let pending = 3;
          function onEnd() {
            if (--pending === 0) {
              if (exitCode === 0) {
                resolve(stream_output);
              } else {
                reject(new Error(
                  'Command existed with code ' + exitCode + ':\n\n' + stream_output + '\n\n' + err_output
                ));
              }
            }
          }
          stream.on('end', () => {
            if (debug) {
              console.log('stdout end');
            }
            onEnd();
          });
          stream.stderr.on('end', () => {
            if (debug) {
              console.log('stderr end');
            }
            onEnd();
          });
          stream.on('exit', code => {
            if (debug) {
              console.log('exit ' + code);
            }
            exitCode = code;
            onEnd();
          });
          if (options.stdin) {
            stream.end(options.stdin);
          }
        });
      });
    });
    this.ready = result.then(null, () => {});
    return result;
  }

  close() {
    const result = this.ready.finally(() => this._connection.end());
    this.ready = Promise.reject(new Error('Cannot interact with a closed connection'));
    return result;
  }
}


src/digital-ocean.js
import Promise from 'promise';
import thenRequest from 'then-request';

const BASE_URL = 'https://api.digitalocean.com/v2';

let timeoutsDisabled = false;
export function _disableTimeouts() {
  timeoutsDisabled = true;
}

export default class DigiatlOcean {
  constructor(token) {
    this._token = token;
    this._runningPolls = 0;
    this._rateLimit = {
      limit: 5000,
      remaining: 5000,
      reset: Math.floor(Date.now() / 1000),
    };
  }
  _timeToRateLimitReset() {
    return Math.max(1000, (this._rateLimit.reset * 1000) - Date.now());
  }
  request(method, path, options) {
    path = path.replace(BASE_URL, '');
    return new Promise((resolve, reject) => {
      const attempt = () => {
        thenRequest(method, BASE_URL + path, {
          ...(options || {}),
          headers: {
            ...(options && options.headers ? options.headers : {}),
            authorization: 'Bearer ' + this._token,
          },
        }).done(res => {
          if (res.headers['ratelimit-limit']) {
            this._rateLimit = {
              limit: res.headers['ratelimit-limit'],
              remaining: res.headers['ratelimit-remaining'],
              reset: res.headers['ratelimit-reset'],
            };
            if (res.statusCode === 429) {
              console.log('Rate limit exceeded, backing off for ' + (this._timeToRateLimitReset() / 1000) + ' seconds');
              if (timeoutsDisabled) return attempt();
              return setTimeout(attempt, this._timeToRateLimitReset());
            }
          }
          resolve(res);
        }, reject);
      };
      attempt();
    });
  }
  requestJSON(method, path, options) {
    return this.request(method, path, options).then(res => {
      return JSON.parse(res.getBody('utf8'));
    });
  }
  getPaged(path, key, filter) {
    return new Promise((resolve, reject) => {
      const result = [];
      const nextPage = (url) => {
        this.requestJSON('get', url).done(response => {
          result.push(...(filter ? response[key].filter(filter) : response[key]));
          if (response.links && response.links.pages && response.links.pages.next) {
            nextPage(response.links.pages.next);
          } else {
            resolve(result);
          }
        }, reject);
      };
      nextPage(path);
    });
  }
  poll(path, isReady, timeout) {
    this._runningPolls++;
    timeout = timeout || (60 * 1000);
    const timeoutEnd = Date.now() + timeout;
    return new Promise((resolve, reject) => {
      const poll = (finalAttempt) => {
        this.requestJSON('get', path).then(result => {
          if (isReady(result)) {
            resolve(result);
          } else if (finalAttempt) {
            reject(new Error('Operation timed out after ' + (timeout / 1000) + ' seconds'));
          } else if (Date.now() > timeoutEnd) {
            poll(true);
          } else {
            // poll less frequently for operations that are typically slower
            // poll less frequently if we are running out of rate limit
            const timeToRateLimitReset = this._timeToRateLimitReset();
            // console.log(
            //   this._rateLimit.remaining + ' remaining of ' + this._rateLimit.limit + ' with reset in ' +
            //   (timeToRateLimitReset / 1000) + ' seconds'
            // );
            const remaining = (this._rateLimit.remaining || 1) / this._runningPolls;
            // console.log('backoff: ' + (remaining / 1000));
            if (timeoutsDisabled) return poll();
            setTimeout(() => poll(), (timeout / 60) + (timeToRateLimitReset / remaining));
          }
        }).done(null, reject);
      };
      poll();
    }).finally(() => this._runningPolls--);
  }
}


src/index.js
import Promise from 'promise';
import sshpk from 'sshpk';
import SSH from './ssh';
import DigitalOcean from './digital-ocean';

let SshImplementation = SSH;
/*eslint-disable */
let getDate = () => (new Date()).toISOString();
/*eslint-enable */

function installNode(ssh, version) {
  return ssh.exec([
    'yum -y update',
    'yum -y groupinstall "Development Tools"',
    'cd ~',
    'wget http://nodejs.org/dist/' + version + '/node-' + version + '-linux-x64.tar.gz',
    'sudo tar --strip-components 1 -xzvf node-v* -C /usr/local',
    'node --version',
  ], {pty: true});
}

class DigitalNavy {
  constructor({token, keypair, name, prepare, maxPoolSize, minPoolSize, workerSize, builderSize}) {
    const client = new DigitalOcean(token);
    this._client = client;
    this._keypair = keypair;
    this._name = name;
    this._workerSize = workerSize || '512mb';
    this._builderSize = builderSize || workerSize || '512mb';

    this._getMaxPoolSize = typeof maxPoolSize === 'function' ? maxPoolSize : () => (maxPoolSize || 2);
    this._getMinPoolSize = typeof minPoolSize === 'function' ? minPoolSize : () => (minPoolSize || 0);
    this._pool = [];
    this._poolSize = 0;
    this._queue = [];

    // Store the ssh key for using when creating droplets
    this._fingerprint = this._getFingerprint();
    this._snapshot = this._buildSnapshot(name, prepare);
    this._updatePool();
    this.ready = Promise.all([this._fingerprint, this._snapshot].concat(this._pool)).then(() => {});
    process.on('beforeExit', () => {
      this._getMinPoolSize = () => 0;
      this._getMaxPoolSize = () => 0;
      this._updatePool();
    });
  }
  // Optionally passed a dirty droplet that is **not** considered part of the current pool
  _updatePool(dirtyDroplet) {
    // Desired size is the number of queued items + the number of in progress items
    let desiredSize = this._queue.length + (this._poolSize - this._pool.length);
    const minSize = this._getMinPoolSize();
    const maxSize = this._getMaxPoolSize();
    if (desiredSize < minSize) {
      desiredSize = minSize;
    }
    if (desiredSize > maxSize) {
      desiredSize = maxSize;
    }
    if (dirtyDroplet) {
      // if our pool is big enough, discard the dirty droplet
      if (desiredSize <= this._poolSize) {
        this._destroyDroplet(dirtyDroplet).done(null, err => {
          console.error('Error Destroying Droplet:');
          console.error(err.stack);
        });
      }
      // if our pool is too small, re-use the dirty droplet
      if (desiredSize > this._poolSize) {
        this._pool.push(this._rebuildDroplet(dirtyDroplet));
        this._poolSize++;
      }
    }
    while (this._poolSize > desiredSize) {
      // take the least ready droplet (i.e. most recently used)
      this._pool.pop().then(droplet => this._destroyDroplet(droplet)).done(null, err => {
        console.error('Error Destroying Droplet:');
        console.error(err.stack);
      });
      this._poolSize--;
    }
    while (this._poolSize < desiredSize) {
      this._pool.push(this._createDroplet({name: 'dnavy-worker-' + this._name, size: this._workerSize}));
      this._poolSize++;
    }
    while (this._queue.length && this._pool.length) {
      const job = this._queue.shift();
      // take the least recently used droplet because it's most likely to already be on and cleaned up
      const droplet = this._pool.shift();
      job(droplet);
    }
  }
  _getFingerprint() {
    const publicKey = sshpk.parseKey(this._keypair.public);
    return this._client.requestJSON('get', '/account/keys/' + publicKey.fingerprint('md5').toString()).then(
      null,
      err => {
        if (err.statusCode !== 404) throw err;
        return this._client.requestJSON('post', '/account/keys', {
          json: {
            name: 'dnavy-' + getDate(),
            public_key: publicKey.toString('ssh'),
          },
        });
      }
    ).then(() => publicKey.fingerprint('md5').toString());
  }

  _buildSnapshot(name, prepare) {
    let d = null;
    return this._client.getPaged(
      '/images',
      'images',
      image => (
        (image.distribution === 'CentOS' && image.public === true && /^\d+\.\d+ x64$/.test(image.name)) ||
        (image.name === name && image.public === false)
      )
    ).then(
      images => images.reduce((current, next) => {
        if (current.name === name) return current;
        if (next.name === name) return next;
        return next.slug > current.slug ? next : current;
      }, images[0])
    ).then(baseImage => {
      if (baseImage.name === name) {
        console.log('reusing existing image ' + name);
        return baseImage;
      }
      return this._createDroplet({
        name: 'dnavy-snapshot-creator-' + name,
        size: this._builderSize,
        image: baseImage,
      }).then(droplet => {
        d = droplet;
        return this._ssh(droplet, ssh => prepare(ssh)).then(
          () => this._shutdownDroplet(droplet)
        ).then(
          () => this._snapshotDroplet(droplet, name)
        );
      }).then(null, err => {
        console.error(err.stack);
      }).finally(() => {
        if (d) return this._destroyDroplet(d);
      });
    });
  }
  _createDroplet({name, size, image}) {
    console.log('create droplet');
    if (!image) {
      image = this._snapshot;
    }
    return Promise.all([
      this._fingerprint,
      image,
    ]).then(([fingerprint, image]) => {
      return this._client.requestJSON('POST', '/droplets', {
        json: {
          name,
          region: 'nyc3',
          size,
          image: image.id,
          ssh_keys: [fingerprint],
          backups: false,
          ipv6: false,
        },
      });
    }).then(result => {
      const id = result.droplet.id;
      return this._client.poll('/droplets/' + id, result => result.droplet.status !== 'new', 4 * 60 * 1000);
    }).then(result => {
      if (result.droplet.status !== 'active') {
        throw new Error(
          'Expected droplet state to go from "new" to "active" but got "' + result.droplet.status + '"'
        );
      }
      return result.droplet;
    });
  }
  _runAction(path, action, options) {
    return this._client.requestJSON(
      'post',
      path + '/actions',
      {json: action},
    ).then(result => {
      const actions = result.actions ? result.actions : [result.action];
      return Promise.all(actions.map(({id}) => {
        return this._client.poll(
          '/actions/' + id,
          result => result.action.status !== 'in-progress',
          options && options.timeout,
        ).then(result => {
          if (result.action.status !== 'completed') {
            throw new Error(action.type + ' did not complete successfully');
          }
        });
      }));
    }).then(() => {});
  }
  _shutdownDroplet(droplet) {
    console.log('shutdown droplet');
    return new Promise((resolve, reject) => {
      const attempt = (attemptNumber) => {
        this._runAction('/droplets/' + droplet.id, {type: 'shutdown'}).then(
          () => {
            return this._client.poll(
              '/droplets/' + droplet.id,
              result => result.droplet.status === 'off'
            );
          },
        ).done(resolve, err => {
          if (attemptNumber < 3) {
            attempt(attemptNumber + 1);
          } else {
            reject(err);
          }
        });
      };
      attempt(0);
    });
  }
  _snapshotDroplet(droplet, name) {
    console.log('snapshot droplet');
    return this._runAction('/droplets/' + droplet.id, {type: 'snapshot', name}, {timeout: 4 * 60 * 1000}).then(() => {
      return new Promise((resolve, reject) => {
        const poll = () => {
          return this._client.getPaged(
            '/images',
            'images',
            image => image.name === name && image.public === false
          ).then(
            images => {
              if (images.length) {
                resolve(images[0]);
              } else {
                poll();
              }
            }
          );
        };
        poll();
      });
    });
  }
  _rebuildDroplet(droplet) {
    console.log('rebuild droplet');
    return this._snapshot.then(snapshot => {
      return this._runAction('/droplets/' + droplet.id, {type: 'rebuild', image: snapshot.id});
    }).then(() => droplet);
  }
  _destroyDroplet(droplet) {
    console.log('destroy droplet');
    return this._client.request('DELETE', '/droplets/' + droplet.id).then(response => {
      if (response.statusCode !== 204) {
        throw new Error('Expected status code 204 while deleting droplet but got ' + response.statusCode);
      }
    });
  }

  _ssh(droplet, options, fn) {
    if (typeof options === 'function') {
      fn = options;
      options = undefined;
    }
    const addresses = droplet.networks.v4.filter(n => n.type === 'public');
    if (addresses.length === 0) {
      throw new Error('Could not find the IP address of droplet');
    }
    const privateKey = sshpk.parsePrivateKey(this._keypair.private);
    const connection = new SshImplementation({
      host: addresses[0].ip_address,
      port: 22,
      username: 'root',
      privateKey: privateKey.toString(),
    }, options);
    return connection.ready.then(() => fn(connection)).finally(() => connection.close());
  }

  run(options, fn) {
    if (typeof options === 'function') {
      fn = options;
      options = {};
    }
    return new Promise((resolve, reject) => {
      this._queue.push(dropletPromise => {
        dropletPromise.then(droplet => {
          return Promise.resolve(null).then(
            () => this._ssh(droplet, {debug: options.debug}, ssh => fn(ssh))
          ).then(
            result => {
              this._poolSize--;
              this._updatePool(droplet);
              return result;
            },
            err => {
              // If anything went wrong, we destroy the droplet just in case the droplet is somehow in a bad state
              this._destroyDroplet(droplet).done(null, err => {
                console.error('Error Destroying Droplet:');
                console.error(err.stack);
              });
              this._poolSize--;
              this._updatePool();
              throw err;
            },
          );
        }).done(resolve, reject);
      });
      // The pool may need to increase in size, or may be big enough to immediately start executing
      this._updatePool();
    });
  }
}

module.exports = DigitalNavy;
DigitalNavy.installNode = installNode;
DigitalNavy._mockSsh = mockSsh => SshImplementation = mockSsh;
DigitalNavy._mockDate = mockDate => getDate = mockDate;
